pdftotree/TreeExtract.py

```diff
@@ -402,42 +402,72 @@ def get_html_others(self, tag: str, box: List[float], page_num: int) -> Element:
-            self.pdf_file, pages=page_num, area=table_str, output_format="json"
+            self.pdf_file, pages=page_num, area=table, output_format="json"
         )
-        if len(table_json) > 0:
-            table_element = self.doc.createElement("table")
-            for i, row in enumerate(table_json[0]["data"]):
-                row_element = self.doc.createElement("tr")
-                table_element.appendChild(row_element)
-                for j, column in enumerate(row):
-                    col_element = self.doc.createElement("td")
-                    row_element.appendChild(col_element)
-                    box = [
-                        column["top"],
-                        column["left"],
-                        column["top"] + column["height"],
-                        column["left"] + column["width"],
-                    ]
-                    elems = get_mentions_within_bbox(box, self.elems[page_num].mentions)
-                    elems.sort(key=cmp_to_key(reading_order))
-                    for elem in elems:
-                        words = self.get_word_boundaries(elem)
-                        for word in words:
-                            top = int(word[1])
-                            left = int(word[2])
-                            bottom = int(word[3])
-                            right = int(word[4])
-                            # escape special HTML chars
-                            text = html.escape(word[0])
-
-                            word_element = self.doc.createElement("span")
-                            col_element.appendChild(word_element)
-                            word_element.setAttribute("class", "ocrx_word")
-                            word_element.setAttribute(
-                                "title", f"bbox {left} {top} {right} {bottom}"
-                            )
-                            word_element.appendChild(self.doc.createTextNode(text))
+        self.log.debug(f"Tabula recognized {len(table_json)} table(s).")
+        if len(table_json) == 0:
+            return None
+        table_element = self.doc.createElement("table")
+        table_element.setAttribute("class", "ocr_table")
+        top = int(table_json[0]["top"])
+        left = int(table_json[0]["left"])
+        bottom = int(table_json[0]["bottom"])
+        right = int(table_json[0]["right"])
+        table_element.setAttribute("title", f"bbox {left} {top} {right} {bottom}")
+        for i, row in enumerate(table_json[0]["data"]):
+            row_element = self.doc.createElement("tr")
+            table_element.appendChild(row_element)
+            for j, cell in enumerate(row):
+                # It is not explicitly stated anywhere but tabula seems to use the cell
+                # bbox to represent that of cell itself rather than that of text inside.
+                # Note: bbox could be [0, 0, 0, 0] if tabula recognizes no text inside.
+                box: List[float] = [
```

Comment:
Can we make this a data class?

Reply:
A data class would be a good idea, but it's only available on Python 3.7 or later.
Instead, I have been actually wondering if we could use [Shapely](https://pypi.org/project/Shapely/) to represent a bbox more broadly in pdftotree.
Shapely comes with R-tree (https://shapely.readthedocs.io/en/latest/manual.html#str-packed-r-tree), which allows fast geometric query like contains or intersect.

Reply:
For now, let's use Shapely for fast geometric query and make an util function that takes `bbox` and returns its string representation for hOCR export.
Once pdftotree drops support for Python 3.6, we can reassess dataclass.

Reply:
I'll address this issue in a separate one.